it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,211 @@
+    this.databaseAdminClient = spanner.getDatabaseAdminClient();
+  }
+
+  public static Builder builder(String testId, String projectId, String region) {
+    return new Builder(testId, projectId, region);
+  }
+
+  synchronized void createInstance(String region) {
+    InstanceInfo instanceInfo =
+        InstanceInfo.newBuilder(InstanceId.of(projectId, instanceId))
+            .setInstanceConfigId(InstanceConfigId.of(projectId, region))
+            .setDisplayName(instanceId)
+            .setNodeCount(1)
+            .build();
+    try {
+      instanceAdminClient.createInstance(instanceInfo).get();
+      LOG.info("Successfully created instance.");
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create instance.", e);
+    }
+  }
+
+  synchronized void createDatabase() {
+    try {
+      databaseAdminClient.createDatabase(instanceId, databaseId, Collections.emptyList()).get();
+      databaseClient = spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+      LOG.info("Successfully created database.");
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create database.", e);
+    }
+  }
+
+  private void checkIsCleanedUp() throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+  }
+
+  @Override
+  public synchronized void createTable(String statement) throws IllegalStateException {
+    checkIsCleanedUp();
+    try {
+      databaseAdminClient
+          .updateDatabaseDdl(instanceId, databaseId, ImmutableList.of(statement), null)
+          .get();
+      LOG.info("Successfully created table.");
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create table.", e);
+    }
+  }
+
+  @Override
+  public synchronized void insertTableRecord(Mutation tableRecord) throws IllegalStateException {
+    insertTableRecords(ImmutableList.of(tableRecord));
+  }
+
+  @Override
+  public synchronized void insertTableRecords(Iterable<Mutation> tableRecords)
+      throws IllegalStateException {
+    checkIsCleanedUp();
+    databaseClient.write(tableRecords);
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, String... columnNames)
+      throws IllegalStateException {
+    return readTableRecords(tableId, ImmutableList.copyOf(columnNames));
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, Iterable<String> columnNames)
+      throws IllegalStateException {
+    checkIsCleanedUp();
+    return databaseClient.singleUse().read(tableId, KeySet.all(), columnNames);
+  }
+
+  @Override
+  public synchronized void cleanupAll() {
+    instanceAdminClient.deleteInstance(instanceId);
```

Comment:
It's not really clear from the documentation how this will respond if the instance doesn't exist (i.e. `createInstance` failed or was never called in the first place). We should have a way of tracking if `createInstance` successfully created the instance. Basically, a boolean field that's false by default, gets set to true if `createInstance` succeeds, and is checked here before making a call.